### ISSUE TITLE
🐛 [Bug] Pet follow movement

### ISSUE THREAD

**Issue by ghost** [vmangos/core#2726]
## 🐛 Bug report
Pets following the player sometimes show strange movement:

1. ~~Player stops, the pet stops, too, but does turn around in an endless loop:~~
https://github.com/user-attachments/assets/4864e989-04ff-4500-bafa-b0c6134dd41b
fixed: d58f8f6b00d95655b1d4a8b1acf641a69eeff55a

2. ~~While the player runs into one direction, the pet sometimes "teleports" a short distance:~~
https://github.com/user-attachments/assets/7213e1da-7600-4990-95e3-ddf6c597952a
most likely a client issue

3. The pet sometimes makes small "jumps" in the air:
https://github.com/user-attachments/assets/8d3991be-98bd-407e-849f-639eb051195f

4. The pet has "hiccups" when swimming:
https://github.com/user-attachments/assets/29f7ed77-e736-4c4b-a57f-288bd003467b

### Expected behavior
The pet should smoothly follow the player. Cmangos Classic comparison:
https://github.com/user-attachments/assets/5b924e54-dd84-4900-a2cb-d1c3c6f314c7

### Steps to reproduce
Create a warlock/hunter and watch the pet movement.

### Version & Environment
Client Version: 1.12.1.5875
Commit Hash: 8ae2cca85af8297a14002577390d665713f6f938

OS Client: Ubuntu 24.04 (Wine)
OS Server: Ubuntu 24.04

**Comment by ghost** [comment]
The first part is fixed with commit d58f8f6b00d95655b1d4a8b1acf641a69eeff55a, but I cannot fix the second part. Played around with TargetedMovementGenerator, tried point movement without pathfinding, checked if spline position is different if the creature is still moving, but did not find the cause of those short "teleports" and "hiccups". Perhaps a kind of movement prediction is needed here.

**Comment by gajet5** [comment]
Could this be related to the problem when pets don't catch up with fleeing mobs?

https://github.com/user-attachments/assets/3ea3a4de-82d9-42d5-a0dc-e17523ed96d6


https://github.com/user-attachments/assets/fbd9795c-28e9-4ad3-9518-4810f95c872d

**Comment by ghost** [comment]
The pet also sometimes makes small "jumps" in the air:
https://github.com/user-attachments/assets/8d3991be-98bd-407e-849f-639eb051195f

Is this only occurring for me?

**Comment by Daribon** [comment]
> The pet also sometimes makes small "jumps" in the air: https://github.com/user-attachments/assets/8d3991be-98bd-407e-849f-639eb051195f
> 
> Is this only occurring for me?

Are you using vanillafixes? 
Because if you are, very high FPS can cause NPCs to teleport into the air while moving. The higher FPS the more it happens.
Limit framerate to 144 fps (or the hz of your monitor provided it's not a 360hz one) if you use vanillafixes.

**Comment by ghost** [comment]
No, I normally use Wine, but I tested this now on 2 PCs and a notebook using Wine with and without VanillaFixes and Windows 10 with and without VanillaFixes. Always the same result. It seems to happen more if the character is fast, e.g. using aura 16618 (Spirit of the Wind).

CMaNGOS uses movement prediction, perhaps something like this can be implemented for VMaNGOS:
https://github.com/cmangos/mangos-classic/blob/master/src/game/MotionGenerators/TargetedMovementGenerator.cpp

**Comment by ghost** [comment]
The small "jumps" are gone when using the maps from Brotalnia's repack (version 23). So there seems to be something wrong with the extractor tools. The short "teleports" still occur from time to time, but I will stick with the maps from the repack for now.

**Comment by ghost** [comment]
> The small "jumps" are gone when using the maps from Brotalnia's repack (version 23). So there seems to be something wrong with the extractor tools. The short "teleports" still occur from time to time, but I will stick with the maps from the repack for now.

No, I was wrong. The small "jumps" still happen from time to time.

**Comment by ghost** [comment]
I tested Cmangos Classic, the issues do not occur there:
https://github.com/user-attachments/assets/5b924e54-dd84-4900-a2cb-d1c3c6f314c7

**Comment by ghost** [comment]
The pet also has "hiccups" when swimming:
https://github.com/user-attachments/assets/29f7ed77-e736-4c4b-a57f-288bd003467b

**Comment by ghost** [comment]
Number 2 mentioned in the issue seems to be caused by the client. It even happens when using point movement without pathfinding. Tested Cmangos Classic a bit more and the small teleports also happen there, especially when running up/down a hill.
